templates/boltzmann/middleware/honeycomb.ts

```diff
@@ -9,77 +10,151 @@ import beeline from 'honeycomb-beeline'
 void `{% endif %}`;
 
+let otelTracer: OtelTracer | null = null;
+
+function isHoneycomb (env: typeof process.env): boolean {
+  return !!env.HONEYCOMB_WRITEKEY
+}
+
+function isOtel (env: typeof process.env): boolean {
+  if (isHoneycomb(env) && env.HONEYCOMB_API_HOST) {
+    return /^grpc:\/\//.test(env.HONEYCOMB_API_HOST)
+  }
+  return false
+}
+
+function getOtelTracer() {
+  if (!otelTracer) {
+    otelTracer = otelTrace.getTracer('boltzmann', '1.0.0')
+  }
+  return otelTracer
+}
+
 function trace ({
   headerSources = ['x-honeycomb-trace', 'x-request-id'],
 } = {}) {
-  if (!process.env.HONEYCOMB_WRITEKEY) {
+  if (!isHoneycomb(process.env)) {
     return (next: Handler) => (context: Context) => next(context)
   }
 
-  const schema = require('honeycomb-beeline/lib/schema')
-  const tracker = require('honeycomb-beeline/lib/async_tracker')
-
-  return function honeycombTrace (next: Handler) {
-    return (context: Context) => {
-      const traceContext = _getTraceContext(context)
-      const trace = beeline.startTrace({
-        [schema.EVENT_TYPE]: 'boltzmann',
-        [schema.PACKAGE_VERSION]: '1.0.0',
-        [schema.TRACE_SPAN_NAME]: `${context.method} ${context.url.pathname}${context.url.search}`,
-        [schema.TRACE_ID_SOURCE]: traceContext.source,
-        'request.host': context.host,
-        'request.original_url': context.url.href,
-        'request.remote_addr': context.remote,
-        'request.method': context.method,
-        'request.scheme': context.url.protocol,
-        'request.path': context.url.pathname,
-        'request.query': context.url.search
-      },
-      traceContext.traceId,
-      traceContext.parentSpanId,
-      traceContext.dataset)
-
-      if (isDev()) {
-        context._honeycombTrace = trace
-      }
-
-      if (traceContext.customContext) {
-        beeline.addContext(traceContext.customContext)
-      }
+  if (isOtel(process.env)) {
+    return function honeycombOtelTrace (next: Handler) {
+      return (context: Context) => {
+        let activeContext = otelContext.active()
+        const tracer = getOtelTracer()
+
+        if (context.headers.traceparent && context.headers.tracestate) {
+          activeContext = otelPropagation.extract(activeContext, {
+            traceparent: context.headers.traceparent,
+            tracestate: context.headers.tracestate
+          })
+        }
+
+        const rootSpan = tracer.startSpan(`${context.method} ${context.url.pathname}${context.url.search}`, {
+          attributes: {
+            'request.host': context.host,
+            'request.original_url': context.url.href,
+            'request.remote_addr': context.remote,
+            'request.method': context.method,
+            'request.scheme': context.url.protocol,
+            'request.path': context.url.pathname,
+            'request.query': context.url.search
+          }
+        }, activeContext)
+
+        otelTrace.setSpan(activeContext, rootSpan)
+
```

Comment:
I might be missing something, but where are we setting the outgoing headers to pass on the two trace header values to downstream services?
that's what the `W3CTraceContextPropagator` is for.
See this example: https://github.com/open-telemetry/opentelemetry-js/issues/2458#issuecomment-913849614

Reply:
Awesome thanks